// vite-express/dist/main.js
'use strict';

var express = require('express');
var fs = require('fs');
var path = require('path');
var pc = require('picocolors');

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

const { NODE_ENV } = process.env;
const Config = {
    mode: (NODE_ENV === "production" ? "production" : "development"),
    inlineViteConfig: undefined,
    transformer: undefined,
};
function info(msg) {
    const timestamp = new Date().toLocaleString("en-US").split(",")[1].trim();
    console.log(`${pc.dim(timestamp)} ${pc.bold(pc.cyan("[vite-express]"))} ${pc.green(msg)}`);
}
function isStaticFilePath(path) {
    return path.match(/(\.\w+$)|@vite|@id|@react-refresh/);
}
function getTransformedHTML(html, req) {
    return Config.transformer ? Config.transformer(html, req) : html;
}
function isRunningViteless() {
    return Config.inlineViteConfig !== undefined;
}
function resolveConfig() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!Config.inlineViteConfig) {
            const { resolveConfig } = yield import('vite');
            return resolveConfig({}, "build");
        }
        const { root = process.cwd(), base = "/", build = { outDir: "dist" }, } = Config.inlineViteConfig;
        return { root, base, build };
    });
}
function getDistPath() {
    return __awaiter(this, void 0, void 0, function* () {
        const config = yield resolveConfig();
        return path.resolve(config.root, config.build.outDir);
    });
}
function serveStatic() {
    return __awaiter(this, void 0, void 0, function* () {
        const distPath = yield getDistPath();
        if (!fs.existsSync(distPath)) {
            info(`${pc.red(`Static files at ${pc.gray(distPath)} not found!`)}`);
            info(`${pc.yellow(`Did you forget to run ${pc.bold(pc.green("vite build"))} command?`)}`);
        }
        else {
            info(`${pc.green(`Serving static files from ${pc.gray(distPath)}`)}`);
        }
        return express.static(distPath, { index: false });
    });
}
const stubMiddleware = (req, res, next) => next();
function injectStaticMiddleware(app, middleware) {
    return __awaiter(this, void 0, void 0, function* () {
        const config = yield resolveConfig();
        const base = config.base || "/";
        app.use(base, middleware);
        const stubMiddlewareLayer = app._router.stack.find((layer) => layer.handle === stubMiddleware);
        if (stubMiddlewareLayer !== undefined) {
            const serveStaticLayer = app._router.stack.pop();
            app._router.stack = app._router.stack.map((layer) => {
                return layer === stubMiddlewareLayer ? serveStaticLayer : layer;
            });
        }
    });
}
function injectViteIndexMiddleware(app, server) {
    return __awaiter(this, void 0, void 0, function* () {
        const config = yield resolveConfig();
        app.get("/*", (req, res, next) => __awaiter(this, void 0, void 0, function* () {
            const template = fs.readFileSync(path.resolve(config.root, "index.html"), "utf8");
            if (isStaticFilePath(req.path))
                next();
            else {
                const html = yield server.transformIndexHtml(req.originalUrl, template);
                res.send(getTransformedHTML(html, req));
            }
        }));
    });
}
function injectIndexMiddleware(app) {
    return __awaiter(this, void 0, void 0, function* () {
        const distPath = yield getDistPath();
        app.use("*", (req, res) => {
            const html = fs.readFileSync(path.resolve(distPath, "index.html"), "utf-8");
            res.send(getTransformedHTML(html, req));
        });
    });
}
function startServer(server) {
    return __awaiter(this, void 0, void 0, function* () {
        const { createServer, mergeConfig } = yield import('vite');
        const config = Config.inlineViteConfig ? yield resolveConfig() : {};
        const vite = yield createServer(mergeConfig(config, {
            clearScreen: false,
            appType: "custom",
            server: { middlewareMode: true },
        }));
        server.on("close", () => vite.close());
        return vite;
    });
}
function config(config) {
    if (config.mode !== undefined)
        Config.mode = config.mode;
    Config.inlineViteConfig = config.inlineViteConfig;
    Config.transformer = config.transformer;
}
function bind(app, server, callback) {
    return __awaiter(this, void 0, void 0, function* () {
        info(`Running in ${pc.yellow(Config.mode)} mode`);
        if (Config.mode === "development") {
            const vite = yield startServer(server);
            yield injectStaticMiddleware(app, vite.middlewares);
            yield injectViteIndexMiddleware(app, vite);
        }
        else {
            if (isRunningViteless()) {
                info(`Custom inline config defined, running in ${pc.yellow("viteless")} mode`);
            }
            yield injectStaticMiddleware(app, yield serveStatic());
            yield injectIndexMiddleware(app);
        }
        callback === null || callback === void 0 ? void 0 : callback();
    });
}
function listen(app, port, callback) {
    const server = app.listen(port, () => bind(app, server, callback));
    return server;
}
function build() {
    return __awaiter(this, void 0, void 0, function* () {
        const { build } = yield import('vite');
        info("Build starting...");
        yield build();
        info("Build completed!");
    });
}
var main = { config, bind, listen, build, static: () => stubMiddleware };

module.exports = main;
